File: PureFunctions/Examples.fs
module PureFunctions.Examples

// Чиста функція: завжди однаковий результат, без змін стану.
let add x y = x + y

// Чиста функція без побічних ефектів та без мутації — вона безпечна для потоків.
let sumList (values: int list) =
    values |> List.sum

// Функція, що може повернути None замість побічного ефекту або аварійного стану.
let tryParseInt (text: string) : int option =
    match System.Int32.TryParse(text) with
    | true, value -> Some value
    | false, _ -> None

// Приклад: функція має сигнатуру, яка не каже про лавку побічних ефектів.
// У F# можна зробити краще, повертаючи option або Result для безпечнішої обробки.
let parsePositiveId (value: string) : int =
    let id = System.Int32.Parse(value)
    if id <= 0 then failwith "ID має бути додатнім"
    id

// Чиста функція трансформації: переворотить рядок без змін глобального стану.
let reverseString (text: string) : string =
    text.ToCharArray()
    |> Array.rev
    |> System.String

// Функція з зовнішнім ефектом, але ми явно сигналізуємо про можливий провал через Result.
let readTextFile (path: string) : Result<string, string> =
    try
        let text = System.IO.File.ReadAllText(path)
        Ok text
    with ex -> Error ex.Message

let exampleUsage () =
    let total = add 3 7
    let sum = sumList [1; 2; 3]
    let maybeValue = tryParseInt "42"
    let fileResult = readTextFile "config.txt"
    let reversed = reverseString "Hello"
    // parsePositiveId може кинути виняток, хоча в сигнатурі це не видно.
    let id = parsePositiveId "10"
    total, sum, maybeValue, fileResult, reversed, id

File: PureFunctions/Examples.cs
using System;
using System.Collections.Generic;

namespace PureFunctions
{
    public static class Examples
    {
        // Чиста функція: результат залежить лише від аргументів.
        public static int Add(int a, int b) => a + b;

        // Чиста функція, без змін стану, без зовнішніх ресурсів.
        public static int SumList(IReadOnlyList<int> values)
        {
            return values.Sum();
        }

        // Функція, що може повернути null замість побічного ефекту або аварійного стану.
        public static int? TryParseInt(string text)
        {
            if (int.TryParse(text, out var value))
                return value;
            return null;
        }

        // Функція, яка може кинути виключення, але сигнатура цього не показує.
        public static int ParsePositiveId(string value)
        {
            if (!int.TryParse(value, out var id))
            {
                throw new FormatException("Неправильний формат ідентифікатора");
            }
            if (id <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(value), "ID має бути додатнім");
            }
            return id;
        }

        // Чиста функція для потоків: створює новий рядок, не змінюючи глобальний стан.
        public static string ReverseString(string input)
        {
            var chars = input.ToCharArray();
            Array.Reverse(chars);
            return new string(chars);
        }

        // Функція, яка може повернути null, хоча сигнатура виглядає як простий string.
        // Це приклад небезпечного API, який не вказує побічні ефекти.
        public static string GetEnvironmentValue(string key)
        {
            return Environment.GetEnvironmentVariable(key);
        }

        public static void ExampleUsage()
        {
            var total = Add(3, 7);
            var sum = SumList(new[] { 1, 2, 3 });
            var maybeValue = TryParseInt("42");
            var envValue = GetEnvironmentValue("PATH");
            var reversed = ReverseString("Hello");
            var id = ParsePositiveId("10");
            Console.WriteLine($"Total: {total}, Sum: {sum}, Maybe: {maybeValue}, Env: {envValue}, Reversed: {reversed}, ID: {id}");
        }
    }
}
